runtime/compiler/env/J9SharedCache.cpp

```diff
@@ -401,25 +406,67 @@ TR_J9SharedCache::matchRAMclassFromROMclass(J9ROMClass * clazz, TR::Compilation
 void *
 TR_J9SharedCache::pointerFromOffsetInSharedCache(void *offset)
    {
-   return (void *) ((UDATA)offset + _cacheStartAddress);
+   UDATA offsetValue = (UDATA)offset;
+   J9SharedClassCacheDescriptor *firstCache = _sharedCacheConfig->cacheDescriptorList->previous;
+   J9SharedClassCacheDescriptor *curCache = firstCache;
+   do
+      {
+      if (offsetValue < curCache->cacheSizeBytes)
+         {
+         return (void *)(offsetValue + (UDATA)curCache->cacheStartAddress);
+         }
+      offsetValue -= curCache->cacheSizeBytes;
+      curCache = curCache->previous;
+      }
+   while (curCache != firstCache);
+   TR_ASSERT_FATAL(false, "Shared cache offset out of bounds");
+   return NULL;
    }
 
 void *
 TR_J9SharedCache::offsetInSharedCacheFromPointer(void *ptr)
    {
-   return (void *) ((UDATA)ptr - _cacheStartAddress);
+   UDATA ptrValue = (UDATA)ptr;
+   UDATA offsetValue = 0;
+   J9SharedClassCacheDescriptor *firstCache = _sharedCacheConfig->cacheDescriptorList->previous;
+   J9SharedClassCacheDescriptor *curCache = firstCache;
+   do
+      {
+      UDATA cacheStart = (UDATA)curCache->cacheStartAddress; // Inclusive
+      UDATA cacheEnd = cacheStart + curCache->cacheSizeBytes; // Exclusive
+      if ((ptrValue >= cacheStart) && (ptrValue < cacheEnd))
+         {
+         return (void *)(ptrValue - cacheStart + offsetValue);
+         }
+      offsetValue += curCache->cacheSizeBytes;
```

Comment:
I'm guessing all the caches have to be the same size right?

Reply:
No, they can differ in size. The code as written should handle this.

Reply:
I see, then 2 clarifications. 

1. `cacheSizeBytes` is the not the soft limit right? Also it's total size, not the amount allocated?
2. `cacheSizeBytes` can't change between runs right?

Reply:
1. Yes in both cases.
2. For the lower (read-only) layers, no. If a lower layer changes size it should invalidate all subsequent layers. The VM should take care of those details.

For the current (write) layer, I'm not sure. Whatever happens today when an SCC changes size should happen here as well. @hangshao0 might be able to answer these questions with more concrete details than I can provide.

Reply:
cacheSizeBytes is not the soft max limit. It is the total cache size and cannot change between runs.